File: src/HolzShots/UI/AboutForm.vb
Imports System.Text
Imports HolzShots.Common
Imports HolzShots.Interop
Imports Microsoft.WindowsAPICodePack.Dialogs

Namespace UI
    Friend Class AboutForm
        Public Shared Property AsIsAboutInstanciated As Boolean
        Private Const YoutubePlaylistPage As String = "https://holz.nu/tunnel"

        Public Sub New()
            InitializeComponent()
            AsIsAboutInstanciated = True
        End Sub

        Private Sub AboutFormFormClosed(sender As Object, e As FormClosedEventArgs) Handles Me.FormClosed
            AsIsAboutInstanciated = False
        End Sub

        Private Sub AboutFormNewLoad(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
            versionLabel.Text = $"{Application.ProductVersion} ({LibraryInformation.VersionDate.ToShortDateString()})"
            timestampLabel.Text = $"2010-{DateTime.Now.Year} {LibraryInformation.PublisherName}"
        End Sub

        Private Shared Sub HolzShotsLinkClicked(ByVal sender As Object, ByVal e As LinkLabelLinkClickedEventArgs) Handles holzShotsLinkLabel.LinkClicked
            LibraryInformation.SiteUrl.SafeProcessStart()
        End Sub

        Private Shared Sub YoutubePLaylistLinkLabelLinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles youtubePlaylistLinkLabel.LinkClicked
            YoutubePlaylistPage.SafeProcessStart()
        End Sub

        Private Sub ShowGfxResourcesLinklabelLinkClicked(sender As Object, e As LinkLabelLinkClickedEventArgs) Handles showGfxResourcesLinklabel.LinkClicked
            ShowGfx(Me)
        End Sub

        Private Shared Sub ShowGfx(parent As IWin32Window)
            Const title = "About Grafics"

            Dim sb As New StringBuilder()
            sb.AppendLine("Marker Icons by:")
            sb.AppendLine("Everaldo Coelho - www.everaldo.com").AppendLine()
            sb.AppendLine("Info icon by:")
            sb.AppendLine("Visual Pharm - www.visualpharm.com").AppendLine()
            sb.AppendLine("Free/remaining icons:")
            sb.AppendLine("www.iconfinder.com and VS2017ImageLibrary")

            If Not TaskDialog.IsPlatformSupported Then
                MessageBox.Show(parent, sb.ToString(), title)
                Return
            End If

            Using td As New TaskDialog()
                td.OwnerWindowHandle = If(parent?.Handle, IntPtr.Zero)
                td.Icon = TaskDialogStandardIcon.Information
                td.Caption = title
                td.InstructionText = "Graphic Resources"
                td.Text = sb.ToString()
                td.StandardButtons = TaskDialogStandardButtons.Close
                td.Show()
            End Using
        End Sub
    End Class
End Namespace

File: src/HolzShots.Windows/Forms/AboutForm.cs
using System.Text;
using System.Windows.Forms;

namespace HolzShots.Windows.Forms;

public partial class AboutForm : Form
{
    private static Lazy<AboutForm> _instance = new(() => new AboutForm());
    public static AboutForm Instance => _instance.Value;

    public AboutForm()
    {
        InitializeComponent();

        VersionLabel.Text = $"{Application.ProductVersion} ({LibraryInformation.VersionDate.ToShortDateString()})";
        ApplicationTitleLabel.Text = LibraryInformation.Name;
    }

    private void LicenseLabel_LinkClicked(object sender, EventArgs e) => IO.HolzShotsPaths.OpenLink(LibraryInformation.LicenseUrl);
    private void HolzShotsLinkLabel_LinkClicked(object sender, EventArgs e) => IO.HolzShotsPaths.OpenLink(LibraryInformation.SiteUrl);
    private void SendFeedbackLink_LinkClicked(object sender, EventArgs e) => IO.HolzShotsPaths.OpenLink(LibraryInformation.IssuesUrl);

    private void ShowGfxResourcesLinklabel_LinkClicked(object sender, EventArgs e)
    {
        const string title = "About Graphics";
        MessageBox.Show(this, AboutDialog.GetGraphicsText(), title);
    }

    private void AboutForm_FormClosed(object sender, FormClosedEventArgs e)
    {
        _instance = new Lazy<AboutForm>(() => new AboutForm());
    }
}

public static class AboutDialog
{
    public static void Show(System.Drawing.Bitmap icon, string version)
    {
        var page = new TaskDialogPage()
        {
            AllowMinimize = true,
            Icon = new TaskDialogIcon(icon),
            Caption = "About " + LibraryInformation.Name,
            Heading = LibraryInformation.Name,
            Text = $"Open Source, {ThisAssembly.Constants.SpdxLicense} licensed screenshot utility that gets out of your way.\n",
            /*
               <a href="website">Website</a>
               <a href="issues">Report Issue</a>
               <a href="license">License</a>
            */
            AllowCancel = false,
            Expander = new TaskDialogExpander()
            {
                Text = "Images/Icons used:\n" + GetGraphicsText(),
            },
            Footnote = new TaskDialogFootnote()
            {
                Text = $"Version: {version}",
            },
            Buttons = [
                TaskDialogButton.Close,
            ],
        };

        // TODO: Maybe check if an update is available?
        /*
        System.Threading.Tasks.Task.Run(async () =>
        {
            await System.Threading.Tasks.Task.Delay(5000);
            page.Footnote.Text += " Update available!";
        });
        */

        TaskDialog.ShowDialog(page, TaskDialogStartupLocation.CenterScreen);
    }

    internal static string GetGraphicsText()
    {
        var sb = new StringBuilder();
        sb.AppendLine("Marker Icons by:");
        sb.AppendLine("Everaldo Coelho - www.everaldo.com").AppendLine();
        sb.AppendLine("Info icon by:");
        sb.AppendLine("Visual Pharm - www.visualpharm.com").AppendLine();
        sb.AppendLine("Free/remaining icons:");
        sb.AppendLine("www.iconfinder.com and VS2017ImageLibrary");
        return sb.ToString();
    }
}
